### PR TITLE
In development

### PR DESCRIPTION
**Pull request by Dvermetten**
This request contains the following updates:
- Modification of the upload procedure by better distinguishing between format and suite
- Fixed issues with user-selected color schemes
- Added better labeling to axes in subplots
- Added download option for figure and table in statistics-tab

### CODE REVIEW

File: R/DataSet.R [hunk 1]
```diff
@@ -94,7 +113,8 @@ DataSet <- function(info, verbose = F, maximization = NULL, format = IOHprofiler
 
     # TODO: add the same sanity checks for TWO_COL format
     if (format != TWO_COL) {
-      maxRT <- sapply(cdat, function(d) d[nrow(d), idxEvals]) %>% set_names(NULL)
+      maxRT <- sapply(rt_unaligned, function(d) d[nrow(d), idxEvals]) %>% set_names(NULL)
```
Comment (wangronin) [inline R/DataSet.R]: Shall we get rid of `magrittr` completely? We also do not need to import it...

File: R/DataSet.R [hunk 2]
```diff
@@ -26,66 +26,85 @@
...
+    FV <- align_function_value(fv_unaligned, format = format)  # function value
 
-    #TODO: check if this works correctly without CDAT-file
-    if (format == TWO_COL)
-      FV <- align_function_value(dat, format = format)  # function value
-    else
-      FV <- align_function_value(cdat, format = format)  # function value
 
     # TODO: remove this and incorporate the parameters aligned by runtimes
     FV[names(FV) != 'FV'] <- NULL
```
Comment (wangronin) [inline R/DataSet.R]: I think we could just use the parameter aligned by FV instead, since it has a higher resolution?

File: R/DataSetList.R [hunk 3]
```diff
@@ -159,6 +163,15 @@ DataSetList <-
       attr(object, 'funcId') <- funcId
       attr(object, 'algId') <- algId
     }
+    
+    suite <- unique(suites)
```
Comment (wangronin) [inline R/DataSetList.R]: Should we make sure that there is only one suite in ONE DataSetList? And we should prevent combining two DataSetLists that have different `suite`?

File: R/plot.R [hunk 4]
```diff
@@ -186,18 +185,25 @@ IOHanalyzer_env$used_colorscheme <- Set3
 #' 
 #' @examples
 #' set_color_scheme("Default")
-set_color_scheme <- function(schemename, path = NULL){
-  if (schemename == "Default") IOHanalyzer_env$used_colorscheme <- Set3
-  else if (schemename == "Variant 1") IOHanalyzer_env$used_colorscheme <- Set2
-  else if (schemename == "Variant 2") IOHanalyzer_env$used_colorscheme <- Set1
+set_color_scheme <- function(schemename, path = NULL, nr_algs = 0){
```
Comment (wangronin) [inline R/plot.R]: I guess those new parameters `nr_algs` are documented / explained?

File: R/plot.R [hunk 5]
```diff
@@ -209,13 +215,14 @@ set_color_scheme <- function(schemename, path = NULL){
...
 get_color_scheme <- function(n){
-  IOHanalyzer_env$used_colorscheme(n)
+  color_palettes(n)
 }
 
 # TODO: incoporate more colors
 color_palettes <- function(ncolor) {
   # TODO: FIX IT!
-  if (ncolor < 5) return(IOHanalyzer_env$used_colorscheme(ncolor))
+  max_colors <- getOption("IOHanalyzer.max_colors", 5)
```
Comment (wangronin) [inline R/plot.R]: Aha, I see the meaning of `max_colors` here

File: R/plotDataSetList.R [hunk 6]
```diff
@@ -639,22 +639,33 @@ Plot.RT.PMF.DataSetList <- function(dsList, ftarget, show.sample = F,
 #' @rdname Plot.RT.Histogram
 #' @export
 Plot.RT.Histogram.DataSetList <- function(dsList, ftarget, plot_mode = 'overlay', use.equal.bins = F){
-
+  if (length(get_funcId(dsList)) != 1 || length(get_dim(dsList)) != 1) {
+    warning("Invalid dataset uploaded. Please ensure the datasetlist contains data
```
Comment (wangronin) [inline R/plotDataSetList.R]: Very good

File: R/plotDataSetList.R [hunk 7]
```diff
@@ -685,26 +696,32 @@ Plot.RT.Histogram.DataSetList <- function(dsList, ftarget, plot_mode = 'overlay'
...
+      disp_y <-  mod(i, n_cols) == 1
+      disp_x <- i > (N - n_cols)
+      disp <- c(disp_x, disp_y)
+      
       p[[i]] %<>%
         add_trace(data = plot_data, x = ~x, y = ~y, width = ~width, type = 'bar',
                   name = algId, text = ~text, hoverinfo = 'text',
                   marker = list(color = rgba_str,
-                                line = list(color = 'rgb(8,48,107)', width = 1.5))) %>%
-        layout(
```
Comment (wangronin) [inline R/plotDataSetList.R]: There is no need to add layout now?

File: R/plotDataSetList.R [hunk 8]
```diff
@@ -988,22 +1018,14 @@ Plot.FV.Histogram.DataSetList <- function(dsList, runtime, plot_mode='overlay',
...
-            x = c(0.5,-0.15), y = c(-0.22, 0.5), showarrow = FALSE
-          )
-        )
+                                line = list(color = 'rgb(8,48,107)', width = 1.5)))
     }
   }
 
   if (plot_mode == 'subplot')
-    p <- subplot(p, nrows = nrows, titleX = F, titleY = F, margin = 0.04)
+    p <- subplot(p, nrows = n_rows, titleX = T, titleY = T, margin = 0.05)
```
Comment (wangronin) [inline R/plotDataSetList.R]: In the future, it is better to calculate / determine the margin automatically

File: R/plotDataSetList.R [hunk 9]
```diff
@@ -1366,10 +1388,10 @@ Plot.RT.Multi_Func.DataSetList <- function(dsList, scale.xlog = F,
   } else if (n_fcts <= 20) {
     n_rows <- ceiling(n_fcts / 3.)
     n_cols <- 3
-  } else if (n_fcts <= 30) {
+  } else {
```
Comment (wangronin) [inline R/plotDataSetList.R]: Shall we make a rule to calculate the number of rows and columns automatically?

File: inst/shiny-server/server/ERTPlot.R [hunk 10]
```diff
@@ -33,22 +33,21 @@ update_ert_per_fct_axis <- observe({
...
-                        scale.xlog = isolate(input$ERTPlot.semilogx),
-                        scale.ylog = isolate(input$ERTPlot.semilogy),
-                        scale.reverse = !attr(data[[1]], 'maximization'))
-    },
-    message = "Creating plot"
+                       show.CI = input$ERTPlot.show.CI,
+                       show.ERT = input$ERTPlot.show.ERT,
+                       show.mean = input$ERTPlot.show.mean,
+                       show.median = input$ERTPlot.show.median,
+                       scale.xlog = input$ERTPlot.semilogx,
```
Comment (wangronin) [inline inst/shiny-server/server/ERTPlot.R]: Also isolate input$ERTPlot.semilogx

File: R/readFiles.R [hunk 11]
```diff
@@ -216,24 +216,41 @@ read_IndexFile_BIOBJ_COCO <- function(fname) {
...
   datafile <- sapply(info, function(item) item$datafile)
-
+  
+  
   format <- lapply(datafile, function(file) {
-    first_line <- scan(file, what = 'character', sep = '\n', n = 1, quiet = T)
+    tryCatch({
+      first_line <- scan(file, what = 'character', sep = '\n', n = 1, quiet = T)
+    }, error = function(e) {
+      stop("Error detecting data files specified in .info, please verify the
```
Comment (wangronin) [inline R/readFiles.R]: good

File: R/readFiles.R [hunk 12]
```diff
@@ -216,24 +216,41 @@ read_IndexFile_BIOBJ_COCO <- function(fname) {
...
+      if (n_col == 2)
+        TWO_COL
+      else
+        IOHprofiler
+    }
     else if (first_line == '%')  # Bi-objective COCO format...
       BIBOJ_COCO
+    else {
+      stop("Error detecting file format of file ", file, "; Please verify
+           the integrity of this file.")
```
Comment (wangronin) [inline R/readFiles.R]: good

File: inst/shiny-server/server/upload.R [hunk 13]
```diff
@@ -143,78 +151,70 @@ observeEvent(selected_folders(), {
...
       )
-
-      DataList$data <- c(DataList$data, new_data)
+      
+      tryCatch(
+        DataList$data <- c(DataList$data, new_data),
+        error = function(e) {
+          print_html(paste('<p style="color:red;">The following error happened', 
+                           'when adding the uploaded data set:</p>'))
+          print_html(paste('<p style="color:red;">', e,
```
Comment (wangronin) [inline inst/shiny-server/server/upload.R]: I need to check those again after merging....

File: R/DataSet.R [hunk 14]
```diff
@@ -26,66 +26,85 @@
...
-    } else if (format == TWO_COL) {
-      datFile <-  file.path(path, paste0(strsplit(filename, '\\.')[[1]][1], '.dat'))
+    
+    suite <- info$suite
+    
+    #Unknown suite, set best guess based on format
+    if (is.null(suite)) {
+      warning("Suite-name not provided in .info-file, taking best guess based on
+              the format of data-files.")
+      suite <- switch(format,
```
[the remark this answers is not in the document]
Comment (Dvermetten) [reply R/DataSet.R]: The guessing of format is still based on the first line of the .dat-file. If it starts with % it is from COCO, if it starts with " it is from IOHexperimenter.

File: R/DataSetList.R (hunk 3 above)
Comment (Dvermetten) [reply R/DataSetList.R]: Yes, that is being done right now, but now it just gives a warning and not an error. I will change it to an error instead

File: R/plotDataSetList.R (hunk 9 above)
Comment (Dvermetten) [reply R/plotDataSetList.R]: Sure, I will add this as a todo

File: R/plotDataSetList.R (hunk 7 above)
Comment (Dvermetten) [reply R/plotDataSetList.R]: Not for the labels, I changed that to use axis labels instead

**Pull request merged**